lib/SourceMapDevToolPlugin.js

```diff
@@ -38,128 +38,126 @@ SourceMapDevToolPlugin.prototype.apply = function(compiler) {
+					allModuleFilenames = ModuleFilenameHelpers.replaceDuplicates(allModuleFilenames, function(filename, i, n) {
+						for(var j = 0; j < n; j++)
+							filename += "*";
+						return filename;
 					});
-					task.modules = modules;
-					task.moduleFilenames = moduleFilenames;
-					return task;
-				}, this).forEach(function(task) {
-					allModules = allModules.concat(task.modules);
-					allModuleFilenames = allModuleFilenames.concat(task.moduleFilenames);
-					tasks.push(task);
+					files.forEach(function(file) {
+						var asset = this.assets[file];
+						var {source, sourceMap} = getSourceAndMap(asset);
+						if(sourceMap) {
+							var {modules, moduleFilenames} = getModulesWithFilenames(sourceMap);
+							moduleFilenames = allModuleFilenames.slice(0, moduleFilenames.length);
+							allModuleFilenames = allModuleFilenames.slice(moduleFilenames.length);
+							modules = modules.slice(0, moduleFilenames.length);
+							allModules = allModules.slice(moduleFilenames.length);
+							sourceMap.sources = moduleFilenames;
+							if(sourceMap.sourcesContent && !options.noSources) {
+								sourceMap.sourcesContent = sourceMap.sourcesContent.map(function(content, i) {
+									return content + "\n\n\n" + ModuleFilenameHelpers.createFooter(modules[i], requestShortener);
+								});
+							} else {
+								sourceMap.sourcesContent = undefined;
+							}
+							sourceMap.sourceRoot = options.sourceRoot || "";
+							var currentSourceMappingURLComment = sourceMappingURLComment;
+							if(currentSourceMappingURLComment !== false && /\.css($|\?)/i.test(file)) {
+								currentSourceMappingURLComment = currentSourceMappingURLComment.replace(/^\n\/\/(.*)$/, "\n/*$1*/");
+							}
+							if(sourceMapFilename) {
+								var filename = file,
+									query = "";
+								var idx = filename.indexOf("?");
+								if(idx >= 0) {
+									query = filename.substr(idx);
+									filename = filename.substr(0, idx);
+								}
+								var sourceMapFile = this.getPath(sourceMapFilename, {
+									chunk: chunk,
+									filename: filename,
+									query: query,
+									basename: basename(filename)
+								});
+								var sourceMapUrl = path.relative(path.dirname(file), sourceMapFile).replace(/\\/g, "/");
+								if(currentSourceMappingURLComment !== false) {
+									delete this.assets[file];
+									this.assets[file] = new ConcatSource(new RawSource(source), currentSourceMappingURLComment.replace(/\[url\]/g, sourceMapUrl));
+								}
+								var sourceMapContent= new RawSource(JSON.stringify(sourceMap));
+								var content = sourceMapContent.source();
+
+								if(!Buffer.isBuffer(content)) {
+									content = new Buffer(content, "utf8"); //eslint-disable-line
+								}
+
+								delete sourceMapContent;
+								var targetPath = compiler.outputFileSystem.join(outputPath, sourceMapFile);
+								compiler.outputFileSystem.writeFile(targetPath, content);
```

Comment:
content stays in memory until writeFile actually writes something which happens later down the line which is one of the reasons we still have a pretty big heap of 5gb